### PR TITLE
Filter members to visible fields only or force by UDA

### PR DESCRIPTION
**Pull request by Domain**

### CODE REVIEW

**Comment by BradleyChatha** [comment]
Nice one, though it does feel slightly hacky, though I guess D doesn't really leave you an awful lot of choice in this case.

(Also sorry for taking so long to get around to this)

A few minor tidbits then I think this'll be fine - there's a few things I can fix up myself after its merged (namely documentation).

File: source/lumars/state.d [hunk 1]
```diff
@@ -2,7 +2,151 @@ module lumars.state;
...
 import taggedalgebraic : visit;
-import std.typecons : Nullable;
+import std.typecons : Nullable, isTuple;
+import std.traits : hasUDA, FieldNameTuple;
+import std.meta : AliasSeq;
+
+struct ShowInLua {}
+struct HideFromLua {}
+struct AsTable {}
+struct AsUserData {}
```
Comment (BradleyChatha) [inline source/lumars/state.d:12]: These seem to be leftovers of some other idea you're brainstorming - they appear to be unused.

File: source/lumars/state.d [hunk 2]
```diff
@@ -845,7 +996,8 @@ struct LuaState
             {
                 if(field == member)
                 {
-                    mixin("ret."~member~"= this.get!(typeof(ret."~member~"))(-1);");
+                    setFieldValue!(member)(ret, get!(FieldType!(T, member))(-1));
+                    //mixin("ret."~member~"= this.get!(typeof(ret."~member~"))(-1);");
```
Comment (BradleyChatha) [inline source/lumars/state.d:1000]: Stray commented code that could be removed

File: source/lumars/state.d [hunk 3]
```diff
@@ -783,7 +927,14 @@ struct LuaState
...
             {
                 idx = cast(int)(i - params.length);
-                if (this.isType!(params[i])(idx))
+                if (params.length != lua_gettop(this.handle))
+                {
+                    static if(i == 0)
+                    {
+                        return getAsStruct!(T, T.fieldNames)(index); 
+                    }
+                }
```
Comment (BradleyChatha) [inline source/lumars/state.d:936]: This seems unrelated - is it another leftover?

**Review by BradleyChatha — CHANGES_REQUESTED** [review]
:raised_hands: Pretty much there now logic wise, just these smaller things.

File: source/lumars/state.d (hunk 3 above)
Comment (Domain) [reply source/lumars/state.d:934]: If I push a Tuple, I cannot get it back without this change. See unittest at line 1300.
